File: data-processing/scala/heremapcontent-difftool/src/main/scala/com/here/platform/data/processing/example/scala/difftool/processor/Compiler.scala
/*
 * Copyright (C) 2017-2021 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.platform.data.processing.example.scala.difftool.processor

import java.nio.charset.StandardCharsets

import com.here.platform.data.processing.blobstore.{Payload, Retriever}
import com.here.platform.data.processing.compiler._
import com.here.platform.data.processing.driver.DriverContext
import com.here.platform.data.processing.spark.partitioner.Partitioner
import com.here.platform.data.processing.example.scala.difftool.processor.LayersDefinitions._
import com.here.schema.rib.v2.topology_geometry._
import com.here.schema.rib.v2.topology_geometry_partition._

import scala.collection.breakOut

object Compiler {
  implicit class RichBoolean(val b: Boolean) extends AnyVal {
    final def option[A](a: => A): Option[A] = if (b) Some(a) else None
  }
}

/**
  * Direct MtoN functional pattern batch processor
  * <p>
  * This batch processor functional pattern is suited for tasks such as processing map tiles.
  * For more information, see the Data Processing Library documentation
  * at https://developer.here.com/olp/documentation.
  */
class Compiler(context: DriverContext)
    extends DirectMToNCompiler[IntermediateData]
    with CompileOut1To1Fn[IntermediateData]
    with LayerDefinitions {
  import Compiler._

  private val previousRetriever: Retriever = context.inRetriever(In.PreviousRib)

  private val currentRetriever: Retriever = context.inRetriever(In.Rib)

  /** The partitioner for the input data
    *
    * If this option is None, the Executor uses a default partitioner.
    * Set this option to apply the given partitioner when querying the
    * input catalogs.
    *
    * @param parallelism The parallelism of the partitioner
    * @return The input optional partitioner with the given parallelism
    *
    */
  override def inPartitioner(parallelism: Int): Option[Partitioner[InKey]] =
    None

  /** The partitioner for the output data
    *
    * If this option is None, the Executor uses a default partitioner.
    * Set this option to apply the given partitioner
    * when querying the output catalogs and producing data to be published.
    *
    * @param parallelism The parallelism of the partitioner
    * @return The output optional partitioner with the given parallelism
    *
    */
  override def outPartitioner(parallelism: Int): Option[Partitioner[OutKey]] =
    None

  /** Calculates the output partition key using the given single input partition key. The catalog and layer
    * values are replaced with the output catalog and output layer respectively; partition ID remains the same.
    * Such mappingFn used because we want to map the current partition key and the corresponding
    * output key from the previous run onto the same out key (in order to perform the diff).
    * Additionally, this data is processed by [[compileOutFn(Key, Iterable)]].
    *
    * @param inKey The input partition being mapped
    * @return The output partitions to which the given input partition maps
    *
    */
  override def mappingFn(inKey: InKey): Iterable[OutKey] =
    Iterable(inKey.copy(catalog = outCatalogId, layer = Out.LayerName))

  /** Calculates the intermediate result from a single input partition. This result is
    * provided together with the input key in [[com.here.platform.data.processing.compiler.direct.CompileOutFn]].
    *
    * @param in The input partition to process
    * @return The value of intermediate data of type `T` for this partition. This value is passed
    *         in [[com.here.platform.data.processing.compiler.direct.CompileOutFn]] to all output keys impacted by the in partition.
    *
    */
  override def compileInFn(in: (InKey, InMeta)): IntermediateData =
    IntermediateData(in.key, in.meta)

  /** Compiles a single output partition.
    *
    * @param outKey The key of the output partition to generate
    * @param intermediate Non-empty [[scala.Iterable]] of intermediate data as provided by compileInFn that is
    *                     required to compile the output partition. This collection has multiple entries with the
    *                     same value if the same value was returned for this outKey by multiple compileInFn
    *                     executions (for different input keys).
    *                     outKeys that do not have any intermediate values assigned from the compileInFn calls get
    *                     deleted by the framework automatically.
    * @return [[scala.Option]] of [[com.here.platform.data.processing.blobstore.Payload]], this
    *         method returns exactly one element with the same values for the output key as passed in outKey.
    *         either the partition content or None if there is no data to be published.
    *
    */
  override def compileOutFn(outKey: OutKey,
                            intermediate: Iterable[IntermediateData]): Option[Payload] = {
    def getSegments(retriever: Retriever)(keyMeta: IntermediateData): Map[String, Segment] = {
      val partition =
        TopologyGeometryPartition.parseFrom(retriever.getPayload(keyMeta.key, keyMeta.meta).content)
      partition.segment.map(x => (x.identifier, x))(breakOut)
    }

    val oldSegments: Map[String, Segment] =
      intermediate
        .find(_.key.catalog == LayersDefinitions.In.PreviousRib)
        .map(getSegments(previousRetriever))
        .getOrElse(Map.empty)

    val newSegments =
      intermediate
        .find(_.key.catalog == LayersDefinitions.In.Rib)
        .map(getSegments(currentRetriever))
        .getOrElse(Map.empty)

    val addedSegments = newSegments.keySet -- oldSegments.keySet
    val removedSegments = oldSegments.keySet -- newSegments.keySet
    val modifiedSegments =
      (oldSegments.keySet intersect newSegments.keySet).filter { segmentId =>
        oldSegments(segmentId).geometry != newSegments(segmentId).geometry
      }

    (addedSegments.nonEmpty || removedSegments.nonEmpty || modifiedSegments.nonEmpty)
      .option {
        def toJSONList(identifiers: Set[String]): String =
          s"[${identifiers.toList.sorted
            .map { x =>
              s""""$x""""
            }
            .mkString(", ")}]"

        val json =
          s"""
             |{
             |  "addedSegments": ${toJSONList(addedSegments)},
             |  "removedSegments": ${toJSONList(removedSegments)},
             |  "modifiedSegments": ${toJSONList(modifiedSegments)}
             |}
          """.stripMargin

        Payload(json.getBytes(StandardCharsets.UTF_8))
      }
  }
}

File: data-processing/java/heremapcontent-difftool/src/main/java/com/here/platform/data/processing/example/java/difftool/processor/Compiler.java
/*
 * Copyright (C) 2017-2021 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.platform.data.processing.example.java.difftool.processor;

import com.google.common.collect.Sets;
import com.google.protobuf.InvalidProtocolBufferException;
import com.here.platform.data.processing.java.Pair;
import com.here.platform.data.processing.java.blobstore.Payload;
import com.here.platform.data.processing.java.blobstore.Retriever;
import com.here.platform.data.processing.java.catalog.partition.Key;
import com.here.platform.data.processing.java.catalog.partition.Meta;
import com.here.platform.data.processing.java.compiler.CompileOut1To1Fn;
import com.here.platform.data.processing.java.compiler.DirectMToNCompiler;
import com.here.platform.data.processing.java.driver.Default;
import com.here.platform.data.processing.java.driver.DriverContext;
import com.here.platform.data.processing.java.spark.partitioner.PartitionerOfKey;
import com.here.schema.rib.v2.TopologyGeometry.Segment;
import com.here.schema.rib.v2.TopologyGeometryPartitionOuterClass.TopologyGeometryPartition;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

/**
 * Shows how to compute the difference between two versions of an input catalog with output to JSON
 *
 * <p>In this example output keys depend only on input keys, so the compiler most suited for this is
 * one of the <a
 * href="https://developer.here.com/documentation/data-processing-library/dev_guide/topics/functional-patterns.html#direct-1n-and-mn-compilers">Direct
 * compilers</a>. The Direct1ToNCompiler does not fit in this case, because in compileOutFn, we may
 * have two intermediate data instances for partition - one for input data and one for the previous
 * version data (if available). Consequently, we use the {@link DirectMToNCompiler}
 *
 * <p>Each output partition should be written to one output layer, once, so we use {@link
 * CompileOut1To1Fn}
 */
public class Compiler
    implements DirectMToNCompiler<IntermediateData>, CompileOut1To1Fn<IntermediateData> {

  /** retriever to get input data */
  private Retriever currentRetriever;

  /** retriever to get previous data */
  private Retriever previousRetriever;

  /**
   * Constructor
   *
   * @param ctx A gate for user to the facilities provided by the framework. Here it is used to get
   *     access to data retrievers that are created using the provided input catalog configuration
   *     and also to get Spark parallelism needed for partitioner
   */
  Compiler(DriverContext ctx) {
    currentRetriever = ctx.inRetriever(Defs.In.RIB_CATALOG);
    previousRetriever = ctx.inRetriever(Defs.In.PREVIOUS_RIB);
  }

  /**
   * The partitioner for the input data
   *
   * <p>As this option is None, the Executor uses a default Hash Partitioner. For this example
   * default should be enough as we do not have references to neighbour partitions
   *
   * @param parallelism The parallelism of the partitioner, this parameter is ignored
   * @return empty, which mean using default Hash Partitioner
   */
  @Override
  public Optional<PartitionerOfKey> inPartitioner(int parallelism) {
    return Optional.empty();
  }

  /**
   * The output partitioner
   *
   * <p>As this option is None, the Executor uses a default Hash Partitioner. For this example
   * default should be enough as we do not have references to neighbour partitions
   *
   * @param parallelism the number of partitioner, this parameter is ignored
   * @return empty, which mean using default Hash Partitioner
   */
  @Override
  public Optional<PartitionerOfKey> outPartitioner(int parallelism) {
    return Optional.empty();
  }

  /**
   * Calculates which output partitions the given single input partition impacts. In this example
   * mapping is 1-to-1, partition of outKey is same as of inKey, with configured Output Layer Id and
   * default output catalog ID
   *
   * @return the output partitions that the given input partition maps to
   */
  @Override
  public Iterable<Key> mappingFn(Key inKey) {
    Key outKey = new Key(Default.OutCatalogId(), Defs.Out.LAYER_NAME, inKey.partition());
    return Collections.singletonList(outKey);
  }

  /**
   * @return layers of the input catalogs that should be queried and provided to the compiler,
   *     grouped by input catalog and identified by catalog id and layer name.
   */
  @Override
  public Map<String, Set<String>> inLayers() {
    return Defs.LayerDefs.inLayers;
  }

  /**
   * @return the output layers compiler intends to populate. As there is only one output catalog, it
   *     is not necessary to specify its identifier
   */
  @Override
  public Set<String> outLayers() {
    return Defs.LayerDefs.outLayers;
  }

  /**
   * Calculates the intermediate result from a single input partition. The result will be provided
   * together with the input key in the {@link
   * com.here.platform.data.processing.java.compiler.CompileOut1To1Fn}.
   *
   * @param in the input partition to process
   * @return the value of intermediate data of type Data for this partition. This value will be
   *     passed in {@link}com.here.platform.data.processing.compiler.CompileOutFn]] to all output
   *     keys impacted by the in partition.
   */
  @Override
  public IntermediateData compileInFn(Pair<Key, Meta> in) {
    return new IntermediateData(in.getKey(), in.getValue());
  }

  /**
   * Compiles a single output partition.
   *
   * @param outKey the key of the output partition to generate
   * @param intermediate non-empty {@link Iterable} of intermediate data as provided by compileInFn
   *     which is required to compile the output partition. This collection will have multiple
   *     entries of the same value in case the same value was returned for this outKey by multiple
   *     compileInFn executions (for different input keys). outKeys which do not have any
   *     intermediate values assigned from the compileInFn calls get deleted automatically.
   * @return {@link Optional} of {@link com.here.platform.data.processing.blobstore.Payload}, this
   *     method is to return exactly one element, with the same values for the output key as passed
   *     in outKey. either the partition content, or None if there is no data to be published.
   */
  @Override
  public Optional<Payload> compileOutFn(Key outKey, Iterable<IntermediateData> intermediate) {

    Map<String, Segment> newSegments =
        getByCatalogId(intermediate, Defs.In.RIB_CATALOG, currentRetriever);
    Map<String, Segment> oldSegments;

    oldSegments = getByCatalogId(intermediate, Defs.In.PREVIOUS_RIB, previousRetriever);

    Set<String> addedSegments = Sets.difference(newSegments.keySet(), oldSegments.keySet());
    Set<String> removedSegments = Sets.difference(oldSegments.keySet(), newSegments.keySet());
    Set<String> modifiedSegments =
        Sets.intersection(oldSegments.keySet(), newSegments.keySet())
            .stream()
            .filter(
                seg ->
                    !newSegments.get(seg).getGeometry().equals(oldSegments.get(seg).getGeometry()))
            .collect(Collectors.toSet());

    if (!modifiedSegments.isEmpty() || !removedSegments.isEmpty() || !addedSegments.isEmpty()) {
      GeometryDiff diff =
          new GeometryDiff(
              toSortedList(modifiedSegments),
              toSortedList(removedSegments),
              toSortedList(addedSegments));
      return Optional.of(new Payload(diff.toByteArray()));
    } else {
      return Optional.empty();
    }
  }

  private static List<String> toSortedList(Set<String> s) {
    List<String> list = new ArrayList<String>(s);
    Collections.sort(list);
    return list;
  }

  private Map<String, Segment> getSegments(Retriever ret, IntermediateData keyMeta) {
    try {
      TopologyGeometryPartition partition =
          TopologyGeometryPartition.parseFrom(ret.getPayload(keyMeta.key, keyMeta.meta).content());
      return partition
          .getSegmentList()
          .stream()
          .collect(Collectors.toMap(Segment::getIdentifier, seg -> seg));
    } catch (InvalidProtocolBufferException e) {
      throw new RuntimeException(e);
    }
  }

  private Map<String, Segment> getByCatalogId(
      Iterable<IntermediateData> intermediateData, String catalogId, Retriever retriever) {
    return StreamSupport.stream(intermediateData.spliterator(), false)
        .filter(data -> data.key.catalog().equals(catalogId))
        .findFirst()
        .map(data -> getSegments(retriever, data))
        .orElse(Collections.emptyMap());
  }
}
